File: StableReleases/Obit-20JAN09/Obit/test/testUVImage.c
#include <stdio.h>
#include <stdlib.h>
#include "ObitSystem.h"
#include "ObitUV.h"
#include "ObitImage.h"
#include "ObitImageUtil.h"
#include "ObitImageMosaic.h"
#include "ObitUVImager.h"

/* program to test UVImage functionality */
int main ( int argc, char **argv )
{
  ObitSystem *mySystem;
  ObitErr *err;
  olong ierr, dim[MAXINFOELEMDIM] = {1,1,1,1,1};
  gchar *AIPSdir[] = {"../AIPSdata/"};
  gchar *FITSdir[] = {"../testIt/"};
  ObitUV *uvdata=NULL;
  ObitImage *fullField=NULL, *outImage=NULL;
  /* data */
  olong inDisk = 1;
  /*gchar *inFile = "UVImageTestIn.uvtab";*/
  gchar *inFile = "WideField20UC.uvtab";
  gchar *sources[] = {"C346R424"};
  olong user=103;
  ObitIOType type = OBIT_IO_FITS;  /* FOR FITS */
  /* ObitIOType type = OBIT_IO_AIPS;  FOR AIPS */
  /* olong Adisk=1, Acno;
     gchar Aname[13] = {"WideField20 "};
     gchar Aclass[7] = {"uvdata"};
     gchar Atype[3] = {"UV"};
     olong  Aseq = 2; */
  olong outDisk  = 1;
  gchar *outFile = "!testUVImageOut.fits";
 /*  olong  masterDisk = 1;
     gchar *masterFile  = "CleanVisMaster.fits";
     olong i;
     ofloat bmaj, bmin, bpa; */
  /* Control */
  gchar *Stokes="I   ";
  ofloat FOV = 0.4167; /* 25/60 = 25' */
  /*gfloat FOV = 0.1; Small for testing */
  ofloat UVRange[] = {0.0,0.0};
  ofloat TimeRange[] = {0.0,10.0};
  ofloat Robust = 0.0;
  ofloat UVTaper[] = {0.0,0.0};
  gboolean doCalSelect = TRUE;
  olong docalib = 2;
  olong gainuse = 2;
  gboolean doFull = TRUE;
  gboolean doBeam = TRUE;
  /* Outlier */
  gchar *Catalog = "NVSSVZ.FIT";
  ofloat OutlierDist = 1.0;  /* Maximum distance to add outlyers (deg) */
  ofloat OutlierFlux = 0.01;/* Minimum estimated outlier flux density (Jy) */
  ofloat OutlierSI   = -1.0; /* Spectral index to estimate flux density */
  olong   OutlierSize = 50;   /* Size of outlyer field */
  gboolean Tr=TRUE, Fl=FALSE;
  gchar *name = "TstName", *class = "Class";
  olong seq=1;
  ObitUVImager *myImager=NULL;

  /* Initialize Obit */
  err = newObitErr();
  ierr = 0;
  mySystem = ObitSystemStartup ("testUVImage", 1, user, 1, AIPSdir, 1, FITSdir, 
				(oint)Tr, (oint)Fl, err);
  ObitErrLog(err); /* show any error messages on err */

  uvdata = newObitUV("input FITS UV data");

  /* setup input */
  /* FOR FITS  */
  ObitUVSetFITS(uvdata, 1000, inDisk,inFile, err);

   /* FOR AIPS
      Adisk = 1;
      Acno = ObitAIPSDirFindCNO(Adisk, user, Aname, Aclass, Atype, Aseq, err);
      if (Acno<0) Obit_log_error(err, OBIT_Error, 
      "Failure looking up input uvdata file");
      ObitUVSetAIPS(uvdata,1000,Adisk,Acno,user,err); */
  /* show any errors */
  if (err->error) ierr = 1;   ObitErrLog(err);  if (ierr!=0) return ierr;
  

 /* Set inputs */
  dim[0] = dim[1] = 1;
  ObitInfoListAlwaysPut(uvdata->info, "doCalSelect", OBIT_bool,   dim, &doCalSelect);
  ObitInfoListAlwaysPut(uvdata->info, "doCalib",     OBIT_long,    dim, &docalib);
  ObitInfoListAlwaysPut(uvdata->info, "gainuse",     OBIT_long,    dim, &gainuse);
  ObitInfoListAlwaysPut(uvdata->info, "doBeam",      OBIT_bool,   dim, &doBeam);
  ObitInfoListAlwaysPut(uvdata->info, "imFileType",  OBIT_long,    dim, &type);
  ObitInfoListAlwaysPut(uvdata->info, "imSeq",       OBIT_long,    dim, &seq);
  ObitInfoListAlwaysPut(uvdata->info, "imDisk",      OBIT_long,    dim, &inDisk);
  ObitInfoListAlwaysPut(uvdata->info, "FOV",         OBIT_float,  dim, &FOV);
  ObitInfoListAlwaysPut(uvdata->info, "doFull",      OBIT_bool,   dim, &doFull);
  ObitInfoListAlwaysPut(uvdata->info, "OutlierDist", OBIT_float,  dim, &OutlierDist);
  ObitInfoListAlwaysPut(uvdata->info, "OutlierFlux", OBIT_float,  dim, &OutlierFlux);
  ObitInfoListAlwaysPut(uvdata->info, "OutlierSI",   OBIT_float,  dim, &OutlierSI);
  ObitInfoListAlwaysPut(uvdata->info, "OutlierSize", OBIT_long,    dim, &OutlierSize);
  ObitInfoListAlwaysPut(uvdata->info, "Robust",      OBIT_float,  dim, &Robust);
  dim[0] = 4;
  ObitInfoListAlwaysPut(uvdata->info, "Stokes",      OBIT_string, dim, Stokes);
  dim[0] = 12;
  ObitInfoListAlwaysPut(uvdata->info, "imName",      OBIT_string, dim, name);
  dim[0] = 6;
  ObitInfoListAlwaysPut(uvdata->info, "imClass",     OBIT_string, dim, class);
  dim[0] = strlen(Catalog);
  ObitInfoListAlwaysPut(uvdata->info, "Catalog",     OBIT_string, dim, Catalog);
  dim[0] = 2;
  ObitInfoListAlwaysPut(uvdata->info, "UVRange",     OBIT_float,  dim, UVRange);
  dim[0] = 2;
  ObitInfoListAlwaysPut(uvdata->info, "UVTaper",     OBIT_float,  dim, UVTaper);
  dim[0] = 2;
  ObitInfoListAlwaysPut(uvdata->info, "timeRange",   OBIT_float,  dim, TimeRange);
  dim[0] = 16;
  ObitInfoListAlwaysPut(uvdata->info, "Sources",      OBIT_string, dim, sources[0]);

  /* Make uvImager object */
  myImager = ObitUVImagerCreate ("Imager Object", uvdata, err);
  if (err->error) Obit_log_error(err, OBIT_Error, "ERROR creating Imager object");
  if (err->error) ierr = 1;   ObitErrLog(err);  if (ierr!=0) return ierr;

  /* Weight */
  ObitUVImagerWeight (myImager, err);
  if (err->error) Obit_log_error(err, OBIT_Error, "ERROR Weighting");
  /* show any errors */
  if (err->error) ierr = 1;   ObitErrLog(err);  if (ierr!=0) return ierr;

  /* Image */
  ObitUVImagerImage (myImager, 0, Fl, Tr, Fl, err);
  /*ObitUVImagerImage (myImager, 0, Tr, Tr, Fl, err);*/
  if (err->error) Obit_log_error(err, OBIT_Error, "ERROR Imaging");
  /* show any errors */
  if (err->error) ierr = 1;   ObitErrLog(err);  if (ierr!=0) return ierr;

  /* Flatten */
  ObitUVImagerFlatten (myImager,  err);
  if (err->error) Obit_log_error(err, OBIT_Error, "ERROR Flattening");
  /* show any errors */
  if (err->error) ierr = 1;   ObitErrLog(err);  if (ierr!=0) return ierr;

  /* Quantize output */
  fullField = ObitImageMosaicGetFullImage (myImager->mosaic, err);
  outImage  = ObitImageUtilQuanFITS(fullField, outFile, outDisk, 0.25, err);
  /* show any errors */
  if (err->error) ierr = 1;   ObitErrLog(err);  if (ierr!=0) return ierr;

  /* Cleanup */
  myImager = ObitUVImagerUnref(myImager);

  /* Shutdown Obit */
  mySystem = ObitSystemShutdown (mySystem);
  
  return 0;
} /* end of main */

File: ObitSystem/Obit/testScripts/testUVImage.py
# Test interferometric imaging script
# The argument, if given, is the data directory, defaults to "../testIt"

import Obit, UV, UVImager, Image, ImageMosaic, OSystem, History, OErr, sys
from OErr import Bomb

if len(sys.argv)>=2:
    dataDir = sys.argv[1]
else:
    dataDir = "../testIt/"

# Init Obit
err=OErr.OErr()
ObitSys=OSystem.OSystem ("Imager", 1, 100, 1, ["../AIPSdata/"], 1, [dataDir], 1, 0, err)
OErr.printErrMsg(err, "Error with Obit startup")

# Allow multiple threads
OSystem.PAllowThreads(2)  # 2 threads

# Files (AIPS)  C346R422    .SPLIT .   1
inDisk = 1
inFile = 'UVImageTestIn.uvtab'
outDisk = 1
outFile = "UVImageTestOut.fits"
masterDisk = 1
masterFile  = 'UVImageTestMaster.fits'

# Imaging parameters
FOV = 0.1/60.0  # Smaller
FOV = 10.0/60.0 # larger 20 cm field
Stokes = 'I'
TimeRange = [0.0,10.0]
UVRange   = [0.0,0.0]
Robust    = 0.0
UVTaper   = [0.0,0.0]


# NVSS catalog
Catalog = 'NVSSVZ.FIT'
OutlierDist = 1.0  # Maximum distance to add outlyers (deg)
OutlierFlux = 0.01 # Minimum estimated outlier flux density (Jy)
OutlierSI   = -1.0 # Spectral index to estimate flux density
OutlierSize = 50   # Size of outlyer field

# If debugging
#Bomb()

# Set data
print "Set data"
inData  = UV.newPFUV("Input data", inFile, inDisk, True, err) 
OErr.printErrMsg(err, "Error initializing")

# Calibration/selection parameters
Input = UVImager.UVCreateImagerInput
# Data selection, cal etc.
Input['InData']      = inData
Input['doCalSelect'] = True
Input['Stokes']      = Stokes
Input['BChan']       = 0
Input['EChan']       = 0
Input['BIF']         = 0
Input['EIF']         = 0
Input['doCalib']     = -1
Input['gainUse']     = 1
Input['flagVer']     = 1
Input['timeRange']   = TimeRange
Input['UVRange']     = UVRange
# Imaging control
Input['doBeam'] = True
Input['Type']   = 0     # FITS
Input['Name']   = 'Image'  # Have to pretend it's AIPS
Input['Class']  = 'Class'
Input['Seq']    = 1
Input['Disk']   = outDisk
Input['FOV']    = 10.0/60.0
Input['FOV']    = FOV # Smaller
Input['xCells'] = 1.0
Input['yCells'] = 1.0
Input['doFull'] = True
Input['NField'] = 0
Input['RAShift'] = [0.0]
Input['DecShift'] = [0.0]
Input['Catalog'] = Catalog
Input['OutlierDist'] = OutlierDist
Input['OutlierFlux'] = OutlierFlux
Input['OutlierSI']   = OutlierSI
Input['OutlierSize'] = OutlierSize
Input['Robust']      =  Robust
Input['UVTaper']     =  UVTaper

print "Create UV Imager"
myImager = UVImager.PUVCreateImager(err, "myMosaic", Input)
OErr.printErrMsg(err, "Error creating image mosaic")

print "Weight uv data"
UVImager.PUVWeight(myImager, err)
OErr.printErrMsg(err, "Error creating image mosaic")

# Make Image
print "Make images"
UVImager.PUVImage(myImager, err, doWeight=False, doBeam=True)
OErr.printErrMsg(err, "Error making images")

# Flatten
print "Flatten images"
UVImager.PUVFlatten(myImager, err)
OErr.printErrMsg(err, "Error flattening image mosaic")

# Full field image for output
outMosaic = UVImager.PGetMosaic(myImager, err)
tmpImage  = ImageMosaic.PGetFullImage (outMosaic, err)

# Do history to scratch image as table
inHistory  = History.History("history", inData.List, err)
outHistory = History.History("history", tmpImage.List, err)
History.PCopyHeader(inHistory, outHistory, err)
# Add this programs history
outHistory.Open(History.READWRITE, err)
outHistory.TimeStamp(" Start Obit "+ObitSys.pgmName,err)
outHistory.WriteRec(-1,ObitSys.pgmName+" inFile = "+inFile,err)
outHistory.WriteRec(-1,ObitSys.pgmName+" FOV = "+str(FOV),err)
outHistory.WriteRec(-1,ObitSys.pgmName+" Stokes = "+Stokes,err)
outHistory.WriteRec(-1,ObitSys.pgmName+" TimeRange = "+str(TimeRange),err)
outHistory.WriteRec(-1,ObitSys.pgmName+" UVRange = "+str(UVRange),err)
outHistory.WriteRec(-1,ObitSys.pgmName+" Robust = "+str(Robust),err)
outHistory.WriteRec(-1,ObitSys.pgmName+" UVTaper = "+str(UVTaper),err)
outHistory.Close(err)
OErr.printErrMsg(err, "Error with history")

# output image
outImage  = Image.newPFImage("Output image", outFile,  outDisk,  False, err)
Image.PClone(tmpImage, outImage, err)   # Same structure etc.

# Copy to quantized integer image with history
print "Write output image"
inHistory  = History.History("history", tmpImage.List, err)
Image.PCopyQuantizeFITS (tmpImage, outImage, err, inHistory=inHistory)

# Compare with master lie [rms diff, max abs diff, max. master]
masterImage  = Image.newPFImage("Master image",   masterFile,  masterDisk,  True, err)
diff = Image.PCompare(outImage, masterImage, err);
print "Comparison, rel. max. residual",diff[1]/diff[0], " rel RMS residual",diff[2]/diff[0]

# Cleanup
ImageMosaic.PZapImage(outMosaic, -1, err)      # delete mosaic images
tmpImage.Zap(err)                              # Full field too
OErr.printErrMsg(err, "Error cleaning up files")

# Say something
print "Imaged",inFile,"to", outFile

# Shutdown Obit
OErr.printErr(err)
OSystem.Shutdown(ObitSys)
